src/main/java/bio/terra/service/snapshotbuilder/query/SourceVariable.java

```diff
@@ -0,0 +1,118 @@
+    this.sourcePointer = sourcePointer;
+    this.isLeftJoin = isLeftJoin;
+  }
+
+  public static SourceVariable forPrimary(SourcePointer sourcePointer) {
+    return new SourceVariable(sourcePointer, false);
+  }
+
+  public void addJoinClause(
+      String joinField, Literal joinFieldOnParent, BinaryFilterVariable.BinaryOperator operator) {
+    var joinClause =
+        new BinaryFilterVariable(makeFieldVariable(joinField), operator, joinFieldOnParent);
+    joinClauses.add(joinClause);
+  }
+
+  public void addJoinClause(
+      String joinField,
+      FieldVariable joinFieldOnParent,
+      BinaryFilterVariable.BinaryOperator operator) {
+    var joinClause =
+        new BinaryFilterVariable(makeFieldVariable(joinField), operator, joinFieldOnParent);
+    joinClauses.add(joinClause);
+  }
+
+  public void addJoinClause(String joinField, FieldVariable joinFieldOnParent) {
+    addJoinClause(joinField, joinFieldOnParent, BinaryFilterVariable.BinaryOperator.EQUALS);
+  }
+
+  public static SourceVariable forJoined(
+      SourcePointer sourcePointer, String joinField, FieldVariable joinFieldOnParent) {
+    return forJoined(sourcePointer, joinField, joinFieldOnParent, false);
+  }
+
+  public static SourceVariable forLeftJoined(
+      SourcePointer sourcePointer, String joinField, FieldVariable joinFieldOnParent) {
+    return forJoined(sourcePointer, joinField, joinFieldOnParent, true);
+  }
+
+  private static SourceVariable forJoined(
+      SourcePointer sourcePointer,
+      String joinField,
+      FieldVariable fieldVariable,
+      boolean isLeftJoin) {
+    var sourceVariable = new SourceVariable(sourcePointer, isLeftJoin);
+    sourceVariable.addJoinClause(joinField, fieldVariable);
+    return sourceVariable;
+  }
+
+  public FieldVariable makeFieldVariable(String fieldName) {
+    FieldPointer fieldPointer = new FieldPointer(sourcePointer, fieldName);
+    return new FieldVariable(fieldPointer, this);
+  }
+
+  public FieldVariable makeFieldVariable(
+      String fieldName,
+      String sqlFunctionWrapper,
+      String alias,
+      String hasCountClause,
+      boolean isDistinct) {
+    FieldPointer fieldPointer =
+        new FieldPointer(sourcePointer, fieldName, sqlFunctionWrapper, hasCountClause);
+    return new FieldVariable(fieldPointer, this, alias, isDistinct);
+  }
+
+  public FieldVariable makeFieldVariable(
+      String fieldName, String sqlFunctionWrapper, String alias, boolean isDistinct) {
+    FieldPointer fieldPointer = new FieldPointer(sourcePointer, fieldName, sqlFunctionWrapper);
+    return new FieldVariable(fieldPointer, this, alias, isDistinct);
+  }
+
+  @Override
+  public String renderSQL(SqlRenderContext context) {
+    String sql = sourcePointer.renderSQL(context);
+    String alias = context.getAlias(this);
+
+    if (alias != null) {
+      sql = new ST("<sql> AS <tableAlias>").add("sql", sql).add("tableAlias", alias).render();
```

Comment:
Do you think this should be just `alias` rather than `tableAlias` now? Since it could be from places other than tables?